azurelinuxagent/ga/exthandlers.py

```diff
@@ -276,6 +288,73 @@ def run(self):
                       message=detailed_msg)
             return
 
+    def get_artifacts_profile(self):
+        try:
+            self.artifacts_profile = self.protocol.get_artifacts_profile()
+        except Exception as e:
+            msg = u"Exception retrieving VmArtifactsProfile blob: {0}".format(ustr(e))
+            detailed_msg = '{0} {1}'.format(msg, traceback.extract_tb(get_traceback(e)))
+            logger.warn(msg)
+
+            add_event(AGENT_NAME,
+                      version=CURRENT_VERSION,
+                      op=WALAEventOperation.ExtensionProcessing,
+                      is_success=False,
+                      message=detailed_msg)
+
+    def process_goal_state(self, instantiation, sequence_number):
+        fabric_goal_state_changed, fast_track_changed = self.determine_what_changed(instantiation, sequence_number)
+        if fabric_goal_state_changed:
+            self.protocol.set_fast_track(False)
+
+        # Fabric goal state takes precedence over VMArtifactsProfile blob goal state
+        # So, process the Fabric goal state first
+        self.handle_ext_handlers(fabric_goal_state_changed)
+        self.last_instantiation = instantiation
+
+        # Now process the VMArtifactsProfile if it changed. Note that just because the VMArtifactsProfile changed
+        # doesn't mean the FastTrack goal state changed
+        if fast_track_changed and self.artifacts_profile is not None:
+            self.process_vm_artifacts_profile_for_fast_track(sequence_number)
+        elif self.protocol.client.get_use_fast_track():
+            # Our last goal state was via VMArtifactsProfile, so reset the extension handlers
+            self.ext_conf = self.protocol.client.get_ext_conf(force=True)
+            self.ext_handlers = self.ext_conf.ext_handlers
+
+    def determine_what_changed(self, instantiation, sequence_number):
```

Comment:
Can you split this method into two methods, with one explicit task each? Something on the lines of `did_fabric_goal_state_change` and `did_new_settings_came_in`? It'll be easier to handle them when testing and debugging as well.

Reply:
Split into two methods